ui/admin/app/components/webhooks/WebhookFormContext.tsx

```diff
@@ -0,0 +1,142 @@
+
+import { Form } from "~/components/ui/form";
+import {
+    WebhookConfirmation,
+    WebhookConfirmationProps,
+} from "~/components/webhooks/WebhookConfirmation";
+import { useAsync } from "~/hooks/useAsync";
+
+export type WebhookFormContextProps = {
+    webhook?: Webhook;
+};
+
+type WebhookFormContextType = {
+    handleSubmit: ReturnType<UseFormHandleSubmit<WebhookFormType>>;
+    isLoading: boolean;
+    error?: unknown;
+    isEdit: boolean;
+    hasToken: boolean;
+    hasSecret: boolean;
+};
+
+const Context = createContext<WebhookFormContextType | null>(null);
+
+const CreateSchema = WebhookSchema;
+const EditSchema = WebhookSchema.omit({ token: true }).extend({
+    secret: z.string(),
+});
+
+export function WebhookFormContextProvider({
+    children,
+    webhook,
+}: WebhookFormContextProps & { children: React.ReactNode }) {
+    const webhookId = webhook?.id;
+
+    const [webhookConfirmation, showWebhookConfirmation] =
+        useState<WebhookConfirmationProps | null>(null);
+
+    const updateWebhook = useAsync(WebhookApiService.updateWebhook, {
+        onSuccess: () => {
+            toast.success("Webhook updated");
+        },
+    });
+    const createWebhook = useAsync(WebhookApiService.createWebhook, {
+        onSuccess: () => {
+            toast.success("Webhook created");
+        },
+    });
+
+    const defaultValues = useMemo<WebhookFormType>(
+        () => ({
+            name: webhook?.name ?? "",
+            description: webhook?.description ?? "",
+            alias: webhook?.alias ?? "",
+            workflow: webhook?.workflow ?? "",
+            headers: webhook?.headers ?? [],
+            secret: "",
+            validationHeader: webhook?.validationHeader ?? "",
+            token: "",
+        }),
+        [webhook]
+    );
+
+    const form = useForm<WebhookFormType>({
+        resolver: zodResolver(webhookId ? EditSchema : CreateSchema),
+        defaultValues,
+    });
+
+    useEffect(() => {
+        form.reset(defaultValues);
+    }, [defaultValues, form]);
+
+    const handleSubmit = form.handleSubmit(async (values) => {
+        const { error, data } = webhookId
+            ? await updateWebhook.executeAsync(webhookId, values)
+            : await createWebhook.executeAsync(values);
+
+        if (error) {
+            if (error instanceof Error) toast.error(error.message);
+            else toast.error("Failed to save webhook");
+
+            return;
+        }
+
+        mutate(WebhookApiService.getWebhooks.key());
+        showWebhookConfirmation({
+            webhook: data,
+            secret: values.secret,
+            token: values.token,
+        });
+    });
+
+    return (
+        <Form {...form}>
+            <Context.Provider
+                value={{
+                    error: updateWebhook.error || createWebhook.error,
+                    isEdit: !!webhookId,
+                    hasSecret: !!webhookId,
```

Comment:
I'm confused about how the webhook ID would be related to whether or not the webhook has a secret set...

Reply:
A webhook does not have to have a secret.

Reply:
discussed with Donnie offline. This has been updated